### PR TITLE
Azure Monitor Exporter - Update LogRecord StateValues

### PR DESCRIPTION
**Pull request by rajkumar-rangaraj**
```text

using var loggerFactory = LoggerFactory.Create(builder => builder
    .AddOpenTelemetry(loggerOptions =>
    {
        loggerOptions.SetResourceBuilder(resourceBuilder);
        loggerOptions.AddAzureMonitorLogExporter(exporterOptions =>
        {
            exporterOptions.ConnectionString = $"InstrumentationKey=Ikey;";
        });
    }));

var logger2 = loggerFactory.CreateLogger("Example");
logger2.LogInformation("Hello from {food} {price}.", "apple", 3.99);

{"name":"Message","time":"2022-02-15T03:28:55.4671461Z","iKey":"66f50587-a821-4a96-8511-fa96e8f28fd7","tags":{"ai.internal.sdkVersion":"dotnet4.8.4420.0:otel1.2.0-rc1:ext1.0.0-alpha.20220214.1","ai.cloud.role":"my-namespace.my-service","ai.cloud.roleInstance":"my-instance"},
"data":{"baseType":"MessageData","baseData":{"message":"Hello from {food} {price}.","severityLevel":"Information","properties":{"food":"apple","price":"3.99"},"ver":2}}}

```

```text

using var loggerFactory = LoggerFactory.Create(builder => builder
    .AddOpenTelemetry(loggerOptions =>
    {
        loggerOptions.SetResourceBuilder(resourceBuilder);
        loggerOptions.AddAzureMonitorLogExporter(exporterOptions =>
        {
            exporterOptions.ConnectionString = $"InstrumentationKey=Ikey;";
        });
        loggerOptions.IncludeFormattedMessage = true;
    }));

var logger2 = loggerFactory.CreateLogger("Example");
logger2.LogInformation("Hello from {food} {price}.", "apple", 3.99);

{"name":"Message","time":"2022-02-15T03:30:12.2025215Z","iKey":"66f50587-a821-4a96-8511-fa96e8f28fd7","tags":{"ai.internal.sdkVersion":"dotnet4.8.4420.0:otel1.2.0-rc1:ext1.0.0-alpha.20220214.1","ai.cloud.role":"my-namespace.my-service","ai.cloud.roleInstance":"my-instance"},
"data":{"baseType":"MessageData","baseData":{"message":"Hello from apple 3.99.","severityLevel":"Information","properties":{"food":"apple","price":"3.99"},"ver":2}}}

```

```text

using var loggerFactory = LoggerFactory.Create(builder => builder
    .AddOpenTelemetry(loggerOptions =>
    {
        loggerOptions.SetResourceBuilder(resourceBuilder);
        loggerOptions.AddAzureMonitorLogExporter(exporterOptions =>
        {
            exporterOptions.ConnectionString = $"InstrumentationKey=Ikey;";
        });
        loggerOptions.IncludeFormattedMessage = true;
        loggerOptions.ParseStateValues = true;
    }));

var logger2 = loggerFactory.CreateLogger("Example");
logger2.LogInformation("Hello from {food} {price}.", "apple", 3.99);

{"name":"Message","time":"2022-02-15T03:31:38.7507314Z","iKey":"66f50587-a821-4a96-8511-fa96e8f28fd7","tags":{"ai.internal.sdkVersion":"dotnet4.8.4420.0:otel1.2.0-rc1:ext1.0.0-alpha.20220214.1","ai.cloud.role":"my-namespace.my-service","ai.cloud.roleInstance":"my-instance"},
"data":{"baseType":"MessageData","baseData":{"message":"Hello from apple 3.99.","severityLevel":"Information","properties":{"food":"apple","price":"3.99"},"ver":2}}}

```

```text

using var loggerFactory = LoggerFactory.Create(builder => builder
    .AddOpenTelemetry(loggerOptions =>
    {
        loggerOptions.SetResourceBuilder(resourceBuilder);
        loggerOptions.AddAzureMonitorLogExporter(exporterOptions =>
        {
            exporterOptions.ConnectionString = $"InstrumentationKey=Ikey;";
        });
        loggerOptions.ParseStateValues = true;
    }));

var logger2 = loggerFactory.CreateLogger("Example");
logger2.LogInformation("Hello from {food} {price}.", "apple", 3.99);

{"name":"Message","time":"2022-02-15T03:34:20.8843815Z","iKey":"66f50587-a821-4a96-8511-fa96e8f28fd7","tags":{"ai.internal.sdkVersion":"dotnet4.8.4420.0:otel1.2.0-rc1:ext1.0.0-alpha.20220214.1","ai.cloud.role":"my-namespace.my-service","ai.cloud.roleInstance":"my-instance"},
"data":{"baseType":"MessageData","baseData":{"message":"Hello from {food} {price}.","severityLevel":"Information","properties":{"food":"apple","price":"3.99"},"ver":2}}}
```

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs [hunk 1]
```diff
@@ -62,12 +63,28 @@ internal static string GetMessage(LogRecord logRecord)
...
             return message;
         }
 
+        internal static void SetLogStateValuesToProperties(LogRecord logRecord, IDictionary<string, string> properties)
+        {
+            if (logRecord.StateValues != null)
+            {
+                for (int i = 0; i < logRecord.StateValues.Count; i++)
+                {
+                    properties.Add(logRecord.StateValues[i].Key, logRecord.StateValues[i].Value.ToString());
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:83]: You may want to exclude {OriginalFormat} else it'll duplicate the info in Message and prop.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/MessageData.cs [hunk 2]
```diff
@@ -10,11 +10,12 @@ internal partial class MessageData
     {
         public MessageData(int version, LogRecord logRecord) : base(version)
         {
-            Message = LogsHelper.GetMessage(logRecord);
-            SeverityLevel = LogsHelper.GetSeverityLevel(logRecord.LogLevel);
-
             Properties = new ChangeTrackingDictionary<string, string>();
             Measurements = new ChangeTrackingDictionary<string, double>();
+
+            Message = LogsHelper.GetMessage(logRecord);
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/MessageData.cs:16]: from a perf standpoint, iterating once for message and iterating again for properties is inefficient. You can iterate just once, get props, and get message from {originalformat} in one go.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs [hunk 3]
```diff
@@ -62,12 +63,28 @@ internal static string GetMessage(LogRecord logRecord)
             {
                 message = logRecord.State.ToString();
             }
-
-            // TODO: Parse logRecord.StateValues to extract originalformat / telemetry properties.
+            else if (logRecord.StateValues != null)
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:66]: i think this should be before the logRecord.State != null condition.

**Comment by rajkumar-rangaraj** [comment]
Thanks @cijothomas!

Examples provided helped arrive with a correct logic for State and StateValues.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/TelemetryExceptionData.cs [hunk 4]
```diff
@@ -15,7 +15,11 @@ internal partial class TelemetryExceptionData
 
         public TelemetryExceptionData(int version, LogRecord logRecord) : base(version)
         {
-            var message = LogsHelper.GetMessage(logRecord);
+            Properties = new ChangeTrackingDictionary<string, string>();
+            Measurements = new ChangeTrackingDictionary<string, double>();
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/TelemetryExceptionData.cs:19]: used anywhere?

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/TelemetryExceptionData.cs:19]: Swagger schema mandates us to use this property. This is not used anywhere, but needs to initialized for serialization to succeed. Swagger update is needed to remove this from all the languages.

**Review by cijothomas — APPROVED** [review]
LGTM. I'd suggest to add unit tests before shipping.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs [hunk 5]
```diff
@@ -122,5 +142,23 @@ internal static SeverityLevel GetSeverityLevel(LogLevel logLevel)
...
+                if (item.Key == "{OriginalFormat}")
+                {
+                    if (!isFormattedMessage)
+                    {
+                        message = item.Value.ToString();
+                    }
+                }
+                else
+                {
+                    properties.Add(item.Key, item.Value.ToString());
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:159]: key might be empty by accident or intentional.eg: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Logs/OpenTelemetryLogger.cs#L93

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs [hunk 6]
```diff
@@ -50,20 +50,40 @@ internal static List<TelemetryItem> OtelToAzureMonitorLogs(Batch<LogRecord> batc
...
-            else if (logRecord.State != null)
+
+            // Both logRecord.State and logRecord.StateValues will not be set at the same time for LogRecord.
+            // Either logRecord.State != null or logRecord.StateValues will be called.
+            if (logRecord.State != null)
+            {
+                if (logRecord.State is IReadOnlyCollection<KeyValuePair<string, object>> stateDictionary)
+                {
+                    ExtractProperties(ref message, properties, stateDictionary, isFormattedMessage);
+                }
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/LogsHelper.cs:70]: something in else part..?

**Review by vishweshbankwar — APPROVED** [review]

**Pull request merged**